// jest-runner-eslint/build/utils/normalizeConfig.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const identity = v => v;

const negate = v => !v;

const asArray = v => typeof v === 'string' ? [v] : v;

const asInt = v => {
  if (typeof v === 'number') {
    return v;
  }

  const int = parseInt(v, 10);

  if (Number.isNaN(int)) {
    throw new Error(`'${v}' cannot be converted to a number`);
  }

  return int;
};

const BASE_CONFIG = {
  cache: {
    default: false
  },
  cacheLocation: {
    default: '.eslintcache'
  },
  config: {
    name: 'configFile',
    default: null
  },
  env: {
    name: 'envs',
    default: [],
    transform: asArray
  },
  ext: {
    name: 'extensions',
    default: ['.js'],
    transform: asArray
  },
  fix: {
    default: false
  },
  fixDryRun: {
    default: false
  },
  format: {
    default: null
  },
  global: {
    name: 'globals',
    default: [],
    transform: asArray
  },
  ignorePath: {
    default: null
  },
  ignorePattern: {
    default: [],
    transform: asArray
  },
  maxWarnings: {
    default: -1,
    transform: asInt
  },
  noEslintrc: {
    name: 'useEslintrc',
    default: false,
    transform: negate
  },
  noIgnore: {
    name: 'ignore',
    default: false,
    transform: negate
  },
  noInlineConfig: {
    name: 'allowInlineConfig',
    default: false,
    transform: negate
  },
  parser: {
    default: null
  },
  parserOptions: {
    default: {}
  },
  plugin: {
    name: 'plugins',
    default: [],
    transform: asArray
  },
  quiet: {
    default: false
  },
  reportUnusedDisableDirectives: {
    default: false
  },
  resolvePluginsRelativeTo: {
    default: undefined
  },
  rules: {
    default: {}
  },
  rulesdir: {
    name: 'rulePaths',
    default: [],
    transform: asArray
  }
};
/* eslint-disable no-param-reassign */

const normalizeCliOptions = rawConfig => Object.keys(BASE_CONFIG).reduce((config, key) => {
  const _BASE_CONFIG$key = BASE_CONFIG[key],
        _BASE_CONFIG$key$name = _BASE_CONFIG$key.name,
        name = _BASE_CONFIG$key$name === void 0 ? key : _BASE_CONFIG$key$name,
        _BASE_CONFIG$key$tran = _BASE_CONFIG$key.transform,
        transform = _BASE_CONFIG$key$tran === void 0 ? identity : _BASE_CONFIG$key$tran,
        defaultValue = _BASE_CONFIG$key.default;
  const value = rawConfig[key] !== undefined ? rawConfig[key] : defaultValue;
  return _objectSpread(_objectSpread({}, config), {}, {
    [name]: transform(value)
  });
}, {});
/* eslint-enable no-param-reassign */


const normalizeConfig = config => {
  return _objectSpread(_objectSpread({}, config), {}, {
    cliOptions: normalizeCliOptions(config.cliOptions || {})
  });
};

module.exports = normalizeConfig;
